internal/psutil.go

```diff
@@ -79,3 +81,86 @@ func WaitForPidToExit(ctx context.Context, queryInterval time.Duration, pid int3
 
 	panic("unreachable code in WaitForPidToExit")
 }
+
+// AverageCPUDeltas returns the average CPU cycles consumed per sampleInterval, as calculated
+// between the time this function is invoked and the time the provided ctx is cancelled.
+func AverageCPUDeltas(ctx context.Context, sampleInterval time.Duration) (*cpu.TimesStat, error) {
+	sampleCount := 0
+	sampleCh := sampleCPUTimes(ctx, sampleInterval)
+
+	firstSample := <-sampleCh
+	if firstSample.Err != nil {
+		return nil, firstSample.Err
+	}
+
+	var lastSample *cpuTimesSample
+	for lastSample = range sampleCh {
+		if lastSample.Err != nil {
+			return nil, lastSample.Err
+		}
+		sampleCount++
+	}
+
+	if lastSample == nil {
+		return nil, errors.New("only got one data point, cannot calculate average")
+	}
+
+	avg := func(first float64, last float64) float64 {
+		return (last - first) / float64(sampleCount)
+	}
+
+	return &cpu.TimesStat{
+		User:      avg(firstSample.User, lastSample.User),
+		System:    avg(firstSample.System, lastSample.System),
+		Idle:      avg(firstSample.Idle, lastSample.Idle),
+		Nice:      avg(firstSample.Nice, lastSample.Nice),
+		Iowait:    avg(firstSample.Iowait, lastSample.Iowait),
+		Irq:       avg(firstSample.Irq, lastSample.Irq),
+		Softirq:   avg(firstSample.Softirq, lastSample.Softirq),
+		Steal:     avg(firstSample.Steal, lastSample.Steal),
+		Guest:     avg(firstSample.Guest, lastSample.Guest),
+		GuestNice: avg(firstSample.GuestNice, lastSample.GuestNice),
+	}, nil
+}
+
+type cpuTimesSample struct {
+	*cpu.TimesStat
+	Err error
+}
+
+func sampleCPUTimes(ctx context.Context, sampleInterval time.Duration) <-chan *cpuTimesSample {
+	returnCh := make(chan *cpuTimesSample)
+	go func() {
+		defer close(returnCh)
+
+		for range time.NewTicker(sampleInterval).C {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			const percpu = false // just give us the aggregate numbers, not each cpu's numbers
+			curTimesList, err := cpu.Times(percpu)
+			if err != nil {
+				returnCh <- &cpuTimesSample{
+					Err: err,
+				}
+				return
+			}
+			curTimes := curTimesList[0] // we set percpu to false, so there's only one stat
+
+			select {
+			case returnCh <- &cpuTimesSample{
+				TimesStat: &curTimes,
+			}:
+			default:
+				returnCh <- &cpuTimesSample{
+					Err: errors.New("caller did not keep up with samples"),
```

Comment:
The reason is that if the caller does not get every sample, their data could be wrong if they are calculating diffs between each data point (something I was doing in a previous iteration of all this). 

However, looking at this again, that is fairly specific to the implementation of the caller rather than this function. I think a better approach is to just include an index with `cpuTimesSample` so then the caller can determine if they missed any data points and decide what to do (if anything). I'll update with that.